Allow discovery configuration of modbus platforms

janiversen opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
Users can now use the new style configuration when configuring the modbus integration. The existing configuration style is kept, allowing users to change gradually. 

Remark the existing configuration style will be removed in a couple of releases.

Example existing configuration:
```yaml
modbus:
  - name: hub1
    type: tcp
    host: IP_ADDRESS
    port: 502

sensor:
  platform: modbus
  registers:
    - name: Sensor1
      hub: hub1
      unit_of_measurement: °C
      slave: 1
      register: 100
```

Same configuration in new style:
```yaml
modbus:
  - name: hub1
    type: tcp
    host: IP_ADDRESS
    port: 502
    sensors:
      - name: Sensor1
        slave: 1
        address: 100
```

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Added possibility to do:
modbus:
binary_sensors: []
sensors: []
switches: []

REMARK the old configuration style is still supported, but will be removed at a later point release, therefore this is not
a breaking change.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ x] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

https://github.com/home-assistant/home-assistant.io/pull/16678

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x ] The code change is tested and works locally.
- [ x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ x] There is no commented out code in this PR.
- [ x] I have followed the [development checklist][dev-checklist]
- [x ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

janiversen committed:
Change modbus configuration to new style.

The old (frozen) configuration is still supported, but when detected a big
warning is issued that it will soon be removed. This allows users to change
their configuration at their pace.

Clean configuration SCHEMAs and move common modbus parts
to MODBUS_SCHEMA (renamed from BASE_SCHEMA).

Add BASE_COMPONENT_SCHEMA to ensure common configuration of components.
All component define e.g. NAME, move these to a common schema.
change components (binary_sensor, sensor, switch) to new config

Add test set for modbus itself (old config and discovery_info).
Add test of devices discovery_info configuration

janiversen committed:
Update discovery_info configuration for sensor.

janiversen committed:
Update discovery_info configuration for switch.

janiversen committed:
Update discovery_info configuration for binary_sensor.

janiversen commented:
@vzahradnik I closed the other PR, because it had label "breaking change", and I modified the commits to make sure it is not a breaking change. Please have a look at this (look in new_config.txt for explanation of the changes).

Once you approve the PR, I will add a PR to update the documentation.

vzahradnik commented on homeassistant/components/modbus/new_config.txt:
```diff
@@ -0,0 +1,45 @@
+		sensors: []
+
+	new parameters:
+		CONF_INPUT_TYPE, optional (CALL_TYPE_COIL) CALL_TYPE_DISCRETE or CALL_TYPE_COIL
```
The old parameter was "register type" with values: holding or input. I guess this is just a typo. Sensor doesn't use coils or discrete inputs.

janiversen replied:
Good catch (Copy/paste is a dangerous tool). Removed the file, since it was merely there so you could see the changes easily.

vzahradnik reviewed (approved):
@janiversen good job! I found one small issue in the `new_config.txt`. It's an easy fix.

janiversen committed:
Review comments.

janiversen commented on homeassistant/components/modbus/__init__.py:
```diff
@@ -69,11 +93,40 @@
-        vol.Required(CONF_SLAVE): cv.positive_int,
+        vol.Optional(CONF_SLAVE): cv.positive_int,
+        vol.Optional(CONF_SCAN_INTERVAL, default=DEFAULT_SCAN_INTERVAL): vol.All(
+            cv.time_period, lambda value: value.total_seconds()
```
I just copied it from the old configuration, but you have a good point. I have changed it to "cv.positive_int", which does the same job.

janiversen commented:
Documentation PR added (#16678) so now it is ready to be merged.

yury-sannikov commented:
@janiversen looks like your Documentation PR link pointing to a wrong repo. [This](https://github.com/home-assistant/home-assistant.io/pull/16678) is the correct one.

janiversen commented:
It was not wrong, I just only added the # number and not the full link. However I have updated it to the full link.

janiversen commented:
Updated to newest dev, no real changes. This PR is ready to be merged.

janiversen committed:
update due to change in core

janiversen committed:
flake8 problem.

janiversen committed:
Correct log message.

janiversen commented:
Updated PR due to changes in dev. These last changes are needed to pass the checks! The last commit was due to coverage failure, and instead of disabling coverage I added new tests.

@MartinHjelmare you motivated me to change the configuration to the new style, and your review comments are very useful (not only to make the integration better, but also for me to learn more about the internals of HA). So please when you have time, please let me hear what should be changed in order to get this merged, thanks in advance.

We have a bunch of open issues and other PR´s depending on this PR (because we are not allowed to change the old config).

vzahradnik commented:
@janiversen I just noticed one thing:
You have defined SCAN_INTERVAL in the BASE_COMPONENT_SCHEMA. However, this attribute is actually in effect only for the COVER and CLIMATE. Other components are now also initialized via `discovery_info`, in which case a default scan interval for the platform is applied.

Modbus Cover and Climate disable polling by the platform:
```py
    @property
    def should_poll(self):
        """Return True if entity has to be polled for state.

        False if entity pushes its state to HA.
        """
        # Handle polling directly in this entity
        return False
```

And handle the update by themselves:
```py
    async def async_added_to_hass(self):
        """Handle entity which will be added."""
        async_track_time_interval(
            self.hass, lambda arg: self._update(), self._scan_interval
        )
```

I stuck to this issue while rebasing the LIGHT against this PR.

janiversen commented:
@vzahradnik good catch, I wonder why my production system works without the should_poll property...anyhow I will add this to all platforms.

vzahradnik commented:
@janiversen also add the call `self.schedule_update_ha_state()` to methods like `self._update()`. See the `Cover` as a reference.

janiversen committed:
add should_poll property.

vzahradnik commented:
@janiversen the `should_poll()` method should be `False` in our case. It is true by default, which means the platform is polling for Modbus changes. However, when we initialize the platform via discovery, there's no way how to pass the `scan_interval` value.

This is the reason, I set polling to False and set up a code to handle polling by itself. It was implemented for `Cover` and `Climate`.

vzahradnik commented:
@janiversen when I look at the current code of `Light`, after we modify the `Switch` to handle the polling by itself, there's very little custom code left in the Light component. But it still can be extended further.

I can prepare a PR against your branch with required changes. Please let me know if you agree.

vzahradnik committed:
Fix polling for Modbus binary sensor

vzahradnik commented:
@janiversen I confirmed that the scan_interval doesn't work in this setup. Binary sensor ignores the value and defaults to 30 seconds. For now, I made a fix for the binary sensor, later today I will fix the rest.

My changes are made against this PR, I will create a PR to your branch.
https://github.com/JSC-electronics/home-assistant-core/tree/fix_polling

vzahradnik committed:
Fix polling for Modbus sensor

janiversen commented:
@vzahradnik There must be a way we can tell core which scan_interval we want to use, it does not sound logical that all platforms need to implement their own timers. I will take a look at the polling code and look for a function to call.

I looked at the "old" code that allowed scan_interval to be defined e.g. at sensor level (not for each single sensor), but I cannot see how the configured scan_interval is passed on to core.

No need to create a PR, I will copy your branch.

vzahradnik commented:
While I implemented this, I looked at other integrations. So far I didn't find a better way. Anyway, I will prepare the change to at least make it work again. If you find a better way, perhaps we could implement this in a separate PR because this is the way how cover and climate work, too.

vzahradnik committed:
Fix polling for Modbus switch

vzahradnik commented:
@janiversen I'm not able to create a PR against your branch. All the changes are implemented here as 3 separate commits:
https://github.com/JSC-electronics/home-assistant-core/tree/fix_polling

Feel free to take a look and you can also pull from there. If you find a better solution, I am open to that. These commits at least fix the current behavior, which is broken with your change.

janiversen commented:
Thanks a lot for helping out. I will amend your changes to this PR later this evening. I wish we had seen this problem a bit earlier, but luckily you found it (and corrected it) before the PR was merged.

We might be able to find a better solution, but that is for another PR later....right now I am eager to submit a bunch of PRs to close most of the open issues (but they all depend on this PR).

vzahradnik commented:
@janiversen and my plan is to finally get Light and Fan merged. I will prepare two very simple PRs to get them reviewed quickly. I remember others suggested some improvements to my Fan code, but I can work on that on another PRs... All these changes depend on this PR, so let's hope it gets merged soon.

janiversen committed:
Fix switch.

janiversen commented:
Added your code (with you as author), but now I have a lot of pytest failures. Including a couple I do not understand (config errors). I will run it locally and make the needed changes.

vzahradnik commented:
Thanks for the update.

janiversen committed:
Fix pytest errors.

janiversen commented:
Finally, there was a couple of "interesting" problems.

@vzahradnik can you please give it a go, and give a final approval, so that the maintainers know it is ready to be merged.

MartinHjelmare commented on homeassistant/components/modbus/__init__.py:
```diff
@@ -69,11 +93,40 @@
-        vol.Required(CONF_CURRENT_TEMP): cv.positive_int,
         vol.Required(CONF_NAME): cv.string,
-        vol.Required(CONF_SLAVE): cv.positive_int,
+        vol.Optional(CONF_SLAVE): cv.positive_int,
```
In the climate schema, slave was required, but in the base schema slave is optional. Does that change matter?

janiversen replied:
First of all, thanks for a long review, as usual most of your requests makes sense (there are one, where I need to think about how to do it).

Slave should only be used when we connect to a modbus device, that have modbus devices "below" it, so it was an error to have it as required.

MartinHjelmare commented on homeassistant/components/modbus/binary_sensor.py:
```diff
@@ -50,11 +71,33 @@
+    #  check for old config:
+    if discovery_info is None:
+        _LOGGER.warning(
+            "Binary_sensor configuration depreciated, will be removed in a future release"
```
```suggestion
            "Binary_sensor configuration is deprecated, will be removed in a future release"
```

MartinHjelmare commented on homeassistant/components/modbus/binary_sensor.py:
```diff
@@ -50,11 +71,33 @@
+            "Binary_sensor configuration depreciated, will be removed in a future release"
+        )
+        discovery_info = {
+            CONF_NAME: "noName",
```
```suggestion
            CONF_NAME: "no name",
```

MartinHjelmare commented on homeassistant/components/modbus/modbus.py:
```diff
@@ -56,9 +62,12 @@ def modbus_setup(
         ):
             if conf_key in conf_hub:
-                load_platform(hass, component, DOMAIN, conf_hub, config)
+                load_platform(hass, component, DOMAIN, conf_hub, conf_hub)
```
The last parameter should be the whole original config.
```suggestion
                load_platform(hass, component, DOMAIN, conf_hub, config)
```

MartinHjelmare commented on homeassistant/components/modbus/sensor.py:
```diff
@@ -99,63 +112,89 @@ def number(value: Any) -> int | float:
+            "Sensor configuration depreciated, will be removed in a future release"
+        )
+        discovery_info = {
+            CONF_NAME: "noName",
```
```suggestion
            CONF_NAME: "no name",
```

MartinHjelmare commented on homeassistant/components/modbus/sensor.py:
```diff
@@ -99,63 +112,89 @@ def number(value: Any) -> int | float:
         )
 
     if not sensors:
         return False
```
Nothing is checking this return value.
```suggestion
        return
```

MartinHjelmare commented on homeassistant/components/modbus/sensor.py:
```diff
@@ -99,63 +112,89 @@ def number(value: Any) -> int | float:
+    #  check for old config:
+    if discovery_info is None:
+        _LOGGER.warning(
+            "Sensor configuration depreciated, will be removed in a future release"
```
```suggestion
            "Sensor configuration is deprecated, will be removed in a future release"
```

MartinHjelmare commented on homeassistant/components/modbus/switch.py:
```diff
@@ -86,15 +93,44 @@ async def async_setup_platform(
+    #  check for old config:
+    if discovery_info is None:
+        _LOGGER.warning(
+            "Switch configuration depreciated, will be removed in a future release"
```
```suggestion
            "Switch configuration is deprecated, will be removed in a future release"
```

MartinHjelmare commented on homeassistant/components/modbus/switch.py:
```diff
@@ -86,15 +93,44 @@ async def async_setup_platform(
+            "Switch configuration depreciated, will be removed in a future release"
+        )
+        discovery_info = {
+            CONF_NAME: "noName",
```
```suggestion
            CONF_NAME: "no name",
```

MartinHjelmare commented on homeassistant/components/modbus/switch.py:
```diff
@@ -108,13 +144,20 @@ def __init__(self, hub: ModbusHub, config: dict[str, Any]):
+        )
+
+    def _update(self):
+        """Please the function above (real update is in the inherited classes."""
```
```suggestion
        """Update the entity state."""
```

MartinHjelmare commented on homeassistant/components/modbus/switch.py:
```diff
@@ -108,13 +144,20 @@ def __init__(self, hub: ModbusHub, config: dict[str, Any]):
+            self.hass, lambda arg: self._update(), self._scan_interval
+        )
+
+    def _update(self):
```
```suggestion
    @abstractmethod
    def _update(self):
```

janiversen replied:
done

MartinHjelmare commented on homeassistant/components/modbus/switch.py:
```diff
@@ -2,6 +2,7 @@
 from __future__ import annotations
 
 from abc import ABC
```
```suggestion
from abc import ABC, abstractmethod
```

MartinHjelmare commented on homeassistant/components/modbus/switch.py:
```diff
@@ -86,15 +93,44 @@ async def async_setup_platform(
+            switches.append(ModbusRegisterSwitch(hub, entry))
     async_add_entities(switches)
 
 
```
We should inherit from `SwitchEntity` not `ToggleEntity` on the line below.

janiversen replied:
done

MartinHjelmare commented on tests/components/modbus/test_init.py:
```diff
@@ -0,0 +1,30 @@
+from homeassistant.components.modbus import number
+
+
+async def test_config_modbus():
+    """Run test for modbus init."""
```
```suggestion
async def test_number_validator():
    """Test number validator."""
```

MartinHjelmare commented on tests/components/modbus/test_modbus.py:
```diff
@@ -0,0 +1,66 @@
+@pytest.mark.parametrize("do_type", ["tcp", "serial"])
+async def test_config_modbus(hass, do_discovery, do_options, do_type):
+    """Run test for modbus."""
+    if do_type == "tcp":
```
One of the points with parametrization is that it allows us to avoid logic in the tests. Can we make the parameters be the config instead so we can remove the if/else logic here?

janiversen replied:
I agree, but it was easier just to type do_type :-) Changed!

janiversen committed:
Update homeassistant/components/modbus/binary_sensor.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

janiversen committed:
Update homeassistant/components/modbus/binary_sensor.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

janiversen committed:
Update homeassistant/components/modbus/modbus.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

janiversen committed:
Update homeassistant/components/modbus/sensor.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

janiversen committed:
Update homeassistant/components/modbus/sensor.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

MartinHjelmare commented:
Our standard is to mark both the deprecation and the invalidation PR as breaking to highlight both in the release notes.

janiversen committed:
Update homeassistant/components/modbus/sensor.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

janiversen committed:
Update homeassistant/components/modbus/switch.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

janiversen committed:
Update homeassistant/components/modbus/switch.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

janiversen committed:
Update homeassistant/components/modbus/switch.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

janiversen committed:
ToogleEntity -> SwitchEntity and add abastract

janiversen committed:
Update homeassistant/components/modbus/switch.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

janiversen committed:
Update tests/components/modbus/test_init.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

janiversen committed:
removed if/else in test.

janiversen commented:
@MartinHjelmare I should have addressed all your suggestion (accepted or implemented all). Please let me know if you have further comments.

janiversen commented:
Rebased to newest dev.

MartinHjelmare commented on tests/components/modbus/test_modbus.py:
```diff
@@ -23,29 +23,30 @@
+    config = {
+        DOMAIN: do_config,
+    }
     if do_options:
```
It would be good to remove this case too and instead extend the parametrization of config.

janiversen replied:
Done

janiversen commented:
I believe that build error is not caused by my changes, but a fluke. Force pushed to see if the error occurs again.

janiversen committed:
Remove other if.

MartinHjelmare reviewed (approved):
Nice!

MartinHjelmare commented:
Please add a breaking change paragraph to the PR description that we can copy to the release notes. Briefly describe the deprecation and what users can do to cope with the change.

janiversen commented:
Updated PR, hope you like the text, otherwise feel free to change it.

MartinHjelmare commented:
Can we add an example of the old style vs the new style of configuration YAML? Then it will be super clear.

janiversen commented:
Of course, updated PR with example.

MartinHjelmare commented:
I don't see the examples in the PR description.

janiversen commented:
hrmffff pressing "update comment" absolutely helps !! my error. It is visible now.

MartinHjelmare commented:
Great!
